File: programs/chronos/src/instructions/revenue_create.rs
use {
    crate::state::*,
    anchor_lang::{prelude::*, solana_program::system_program},
    std::mem::size_of,
};

#[derive(Accounts)]
#[instruction(bump: u8)]
pub struct RevenueCreate<'info> {
    #[account(
        seeds = [
            SEED_DAEMON, 
            daemon.owner.as_ref()
        ],
        bump = daemon.bump,
        owner = crate::ID
    )]
    pub daemon: Account<'info, Daemon>,

    #[account(
        init,
        seeds = [
            SEED_REVENUE, 
            daemon.key().as_ref()
        ],
        bump = bump,
        payer = signer,
        space = 8 + size_of::<Revenue>(),
    )]
    pub revenue: Account<'info, Revenue>,

    #[account(mut)]
    pub signer: Signer<'info>,

    #[account(address = system_program::ID)]
    pub system_program: Program<'info, System>,
}

pub fn handler(ctx: Context<RevenueCreate>, bump: u8) -> ProgramResult {
    // Get accounts.
    let daemon = &mut ctx.accounts.daemon;
    let revenue = &mut ctx.accounts.revenue;

    // Initialize revenue account.
    revenue.daemon = daemon.key();
    revenue.balance = 0;
    revenue.bump = bump;

    Ok(())
}

File: programs/chronos/client/src/instruction/revenueCreate.ts
import { Program } from "@project-serum/anchor";
import {
  PublicKey,
  SystemProgram,
  TransactionInstruction,
} from "@solana/web3.js";
import { Chronos } from "../idl";
import { Account } from "../account";

export type RevenueCreateArgs = {
  daemon: PublicKey;
  signer: PublicKey;
};

export class RevenueCreate {
  private account: Account;
  private chronos: Program<Chronos>;

  constructor(account: Account, chronos: Program<Chronos>) {
    this.account = account;
    this.chronos = chronos;
  }

  public async revenueCreate({
    daemon,
    signer,
  }: RevenueCreateArgs): Promise<TransactionInstruction> {
    const revenuePDA = await this.account.revenue.pda(daemon);
    return this.chronos.instruction.revenueCreate(revenuePDA.bump, {
      accounts: {
        daemon: daemon,
        revenue: revenuePDA.address,
        signer: signer,
        systemProgram: SystemProgram.programId,
      },
    });
  }
}
